### DISCUSSION TITLE
dev.morphia.Datastore Thread Safety

### DISCUSSION THREAD

**Discussion by AmitKuGarg** [MorphiaOrg/morphia#2802] [General]
**Describe the bug**
When dev.morphia.Datastore shared across thread, I get Two entities have been mapped using the same discriminator value.

**To Reproduce**
1. Create a sample collection and create the datastore.
```java
Datastore datastore = Morphia.createDatastore(mongo, dbName);
datastore.ensureIndexes();
if(this.packages != null){
   for (String packageString : packages) {
      datastore.getMapper().map(packageString);
   }
}
```

3. Create a runnable class and pass the data store and id parameter. In run method load the data from the collection using id and datastore. 
```java
class Task implements Runnable {
    private String id;
    private Datastore mongoDatastore;

    Task(String id, Datastore mongoDatastore) {
        this.id = id;
        this.mongoDatastore = mongoDatastore;
    }

    @Override
    public void run() {

        Sample result = mongoDatastore.find(Sample.class).filter(Filters.eq("_id", id)).first();        
        if(result == null){
            System.out.println("Result is null");
        }else {
            System.out.println(result.toString());
        }
    }
}
```

4. Write main method, create ExecutorService and try to load different documents.
```java
ExecutorService executor = Executors.newFixedThreadPool(10);
executor.execute(new Task("ABC12345", mongoDatastore));
executor.execute(new Task("ABC123456", mongoDatastore));
```

I get Two entities have been mapped using the same discriminator value.

Try two scenario
1. When ids are valid id from collection
2. Try some invalid ids where collection return null.

**Comment by AmitKuGarg** [comment]
I am creating data source like below
com.mongodb.ConnectionString con = new com.mongodb.ConnectionString("end point");
com.mongodb.client.MongoClient mongo =  com.mongodb.client.MongoClients.create(connectionString);
dev.morphia.Datastore datastore = Morphia.createDatastore(mongo, "dbName");

**Comment by evanchooly** [comment]
What version are you on?  2.4.10 should have all that cleaned up but it should be noted that `Mapper` and `Datastore` are not thread safe.

**Comment by AmitKuGarg** [comment]
I am using morphia-core - 2.4.5 & mongodb-driver-sync - 4.10.2.
 As you mentioned that Mapper and DataStore are not thread safe, does it mean that I need to create a new datastore/mapper for each thread.  

I want to use Morphia datastore in my web application (many concurrent user/request to pull/save data) and creating new data store for each request might waste a lot of resource. Could you please advise best way to do that.

 
Thank you in advance.

**Comment by AmitKuGarg** [comment]
@evanchooly Could you please advise, if I am using Morphia in a wrong way. Any kind of help, I will highly apricate. I need to make some decision.

Again, many thanks in advance.

**Comment by evanchooly** [comment]
The mapping process is the sticking point and I've been working on making it thread safer.  You're hitting a corner case that isn't necessarily destructive and a later version has ensmartened that check to avoid false positive warnings.  So I would suggest bumping to 2.4.10 (and I'm about push 2.4.11 to fix a query document shape bug) to clear up those warnings.

**Comment by AmitKuGarg** [comment]
I tested with version 2.4.11 and issue is still there. If I run the same test multiple time, issue appears.
